# This is a GitHub Repository of code used for the publication titled: Genome-resolved metaproteomics decodes the microbial and viral contributions to coupled carbon and nitrogen cycling in river sediments.


The different files in this repository have all the necessary steps that are required to re-create the output that we used to write this manuscript. In each file, you will either find specific commands on what exactly was run, and how to get input files for them if they are part of a pipeline, OR, they will contain commands as well as the input files that were used (e.g., 2.1..., 2.2...3.1...) where each whole number corresponds to the main command and the decimals represent the additional files provided.


## Manuscript abstract:

Rivers have a significant role in global carbon and nitrogen cycles, serving as a nexus for nutrient transport between terrestrial and marine ecosystems. Although rivers have a small global surface area, they contribute substantially to global greenhouse gas emissions through microbially mediated processes within the river hyporheic zone. Despite this importance, microbial roles in these climatically relevant systems are mostly inferred from 16S rRNA amplicon surveys, which are not sufficiently resolved to inform biogeochemical models. To begin to survey the metabolic potential and gene expression underpinning carbon and nitrogen biogeochemical cycling in rivers sediments, we collected integrated dataset of over thirty metagenomes, metaproteomes, and paired metabolomes. We reconstructed over 500 microbial metagenome assembled genomes (MAGs), which we dereplicated into 55 unique genomes spanning 12 bacterial and archaeal phyla. We also reconstructed 188 viral genomic contigs >10kb which were dereplicated into 111 viral MAGs. As a result of integrating gene expression data with geochemical and metabolite data, we formulated a conceptual model that uncovers new roles for microorganisms in organic matter decomposition, carbon sequestration, nitrogen mineralization, nitrification, and denitrification. Integrated through shared resource pools of ammonium, carbon dioxide, and inorganic nitrogen we show how these metabolic pathways could ultimately contribute to carbon dioxide and nitrous oxide fluxes from hyporheic sediments. Further, by linking viral genomes to these active microbial hosts, we provide some of the first insights into viral modulation of river sediment carbon and nitrogen cycling.

## Author list:

Josué A. Rodríguez-Ramos1, Mikayla A. Borton1,2, Bridget B. McGivern1, Garrett J. Smith3, Lindsey M. Solden1, Michael Shaffer1, Rebecca A. Daly1, Samuel O. Purvine4, Carrie D. Nicora2, Elizabeth K. Eder4, Mary Lipton4, David W. Hoyt4, James C. Stegen2, and Kelly C. Wrighton1* 

1. Department of Soil and Crop Sciences, Colorado State University, Fort Collins, CO, USA
2. Biological Sciences Division, Pacific Northwest National Laboratory, Richland, WA, USA
3. Department of Microbiology, Radboud University, Nijmegen, Netherlands
4. Environmental and Biological Sciences Directorate, Pacific Northwest National Laboratory, Richland, WA, USA

\*Corresponding author: Kelly Wrighton
